Rubberduck.Parsing/ComReflection/ComVariant.cs

```diff
@@ -59,12 +88,79 @@ private struct Variant
 
         public ComVariant(IntPtr variant)
         {
-            Value = Marshal.GetObjectForNativeVariant(variant);
             var members = Marshal.PtrToStructure<Variant>(variant);
             VariantType = (VarEnum)members.vt;
-            if (Value == null && VariantType == VarEnum.VT_BSTR)
+
+            // Note that these are technically flags, but it they are VT_BYREF or VT_ARRAY, the data area will always be a pointer.
+            // ReSharper disable once SwitchStatementMissingSomeCases
+            switch ((VarEnum)members.vt)
             {
-                Value = string.Empty;
+                case VarEnum.VT_VOID:
+                case VarEnum.VT_EMPTY:
+                case VarEnum.VT_NULL:
+                case VarEnum.VT_ERROR:
+                    Value = null;
+                    break;
+                case VarEnum.VT_I2:
+                    Value = members.sVal;
+                    break;
+                case VarEnum.VT_HRESULT:
+                case VarEnum.VT_I4:
+                    Value = members.iVal;
+                    break;
+                case VarEnum.VT_R4:
+                    Value = members.fltVal;
+                    break;
+                case VarEnum.VT_R8:
+                    Value = members.dblVal;
+                    break;
+                case VarEnum.VT_CY:
+                    Value = members.dblVal * 10000;
```

Comment:
This feels iffy to me. 1, it's not clear what the resulting type will be -- if it get converted to a `double` then that would be bad because it's supposed to be precise. 2, I would rather use a proper conversion function -- perhaps `Var***FromCy` in the oleaut32 function. IDK if `VarDecFromCy` would be easily converted to the `decimal` data type but that would be the most faithful conversion, I think.

Reply:
IIR there's an Interop type for `CoCurrency` or something like that, but I can't imagine what it's doing differently. My understanding is that VT_CY is simply a VT_R8 scaled by 10000. I *do* know that it shouldn't be `decimal`.

Reply:
Oh, crap - I completely misremembered that. I'll see if I can dig up a managed equivalent and if not import `VarDecFromCy`.